### PR TITLE
MM-55208 allow search prefixes

### LINKED ISSUE #25354: tilde ~ and at @ cause searches to return fewer results than expected
[https://github.com/mattermost/mattermost/issues/25223](https://github.com/mattermost/mattermost/issues/25223|smart-link) 

#### Summary

The `~` tilde and `@` at are the sigils to indicate channel names and usernames respectively throughout most of the interface. However, in the Search field, adding `~` to a channel name in the `in:` field will cause nothing to match. Adding `@` to a username in the `from:` field will cause nothing to match.

#### Steps to reproduce

- create a new channel, `~brand-new-channel`
- write `hello` to the new channel
- in the search field, `in:~brand-new-channel from:seth-arnold hello` --> `No results for "in:~brand-new-channel from:seth-arnold hello"`
- in the search field, `in:~brand-new-channel from:@seth-arnold hello` --> `No results for "in:~brand-new-channel from:@seth-arnold hello"`
- in the search field, `in:brand-new-channel from:seth-arnold hello` --> The new message that was sent a few seconds ago shows up in the search results pane.

#### Expected behavior

I expect prepending a single `~` on channel names or `@` on user names to not modify the search results. To my mind, they are part of the object identifier and when they are forbidden in the Search field it is very confusing.

#### Observed behavior

- `No results for "in:~brand-new-channel from:seth-arnold hello"`
- `No results for "in:~brand-new-channel from:@seth-arnold hello"`

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-55208

### PR DESCRIPTION
**Pull request by Willyfrog**
#### Summary

Allow using @ and ~ at the `in:` search modifier without affecting search results

Additionally, as suggested by @asaadmahmood , adding the tilde doesn't break the suggestion box. Suggested in a stale community PR: https://github.com/mattermost/mattermost/pull/25594/

Fixes #25354

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55208

#### Screenshots

CleanShot 2024-09-10 at 13 14 53

#### Release Note
```release-note
 Allow using @ and ~ at the `in:` search modifier without affecting search results
```

### CODE REVIEW

**Review by asaadmahmood — APPROVED** [review]
LGTM

**Review by sbishel — APPROVED** [review]
Looks good. Minor suggestion on improving unit test.

**Review by harshilsharma63 — APPROVED** [review]

File: server/channels/app/post_test.go [hunk 1]
```diff
@@ -1768,6 +1768,48 @@ func TestSearchPostsForUser(t *testing.T) {
...
+		assert.Nil(t, err)
+		assert.Equal(t, []string{
+			posts[6].Id,
+			posts[5].Id,
+			posts[4].Id,
+			posts[3].Id,
+			posts[2].Id,
+			posts[1].Id,
+			posts[0].Id,
+		}, results.Order)
```
[the remark this answers is not in the document]
Comment (Willyfrog) [reply server/channels/app/post_test.go]: Thanks! I wasn't too happy about the test. Implemented

**Comment by lindy65** [comment]
@Willyfrog - can you take a look at the failing checks please? I think some of them are the reason a spinwick won't build. Thanks :)

**Review by lindy65 — APPROVED** [review]
Thanks @Willyfrog - this is fixed

**Pull request merged**